File: Data Structures & Algorithms/validate-parentheses/submission-5.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for i in range(len(s)):
            if len(stack) == 0:
                if s[i] == "]" or s[i] == "}" or s[i] == ")":
                    return False
            if s[i] == "(" or s[i] == "[" or s[i] == "{":
                stack.append(s[i])
            if s[i] == ")":
                if stack[-1] == "(":
                    stack.pop()
                else:
                    return False
            if s[i] == "]":
                if stack[-1] == "[":
                    stack.pop()
                else:
                    return False
            if s[i] == "}":
                if stack[-1] == "{":
                    stack.pop()
                else:
                    return False
        
        return len(stack) == 0
        

File: Data Structures & Algorithms/minimum-stack/submission-5.cpp
class MinStack {
private: 
    long min;
    std::stack<long> stack;

public:
    MinStack() {
        
    }
    
    void push(int val) {
        if(stack.empty()){
            stack.push(0);
            min = val;
        }
        else {
            stack.push(val-min);
            if (val < min){
                min = val;
            }
        }
    }
    
    void pop() {
        if(stack.empty()){
            return;
        }
        long pop = stack.top();
        stack.pop();
        if (pop<0){
            min = min - pop;
        }

    }
    
    int top() {
        long top = 0;
        top = stack.top();
        if(top> 0){
            return min+top;
        }
        else{
            return int(min);
        }
    }
    
    int getMin() {
        return int(min);
    }
};
